### PR TITLE
Create a SpatialLocalizer registration system to support multiple types of localization from the same compiled app

### PR DESCRIPTION
**Pull request by matthejo**
## Changes
- Fixes: # .

### CODE REVIEW

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemParticipant.cs [hunk 1]
```diff
@@ -6,267 +6,98 @@
 using Microsoft.MixedReality.Toolkit.Utilities;
 using System;
 using System.IO;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using UnityEngine;
+using UnityEngine.XR.WSA;
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemParticipant.cs]: Why is this needed/does it build for ios/android?

File: Assets/MixedRealityToolkit.Extensions/Sharing/SpatialAlignment/Common/SpatialCoordinateServiceBase.cs [hunk 2]
```diff
@@ -65,15 +65,20 @@ protected override void OnManagedDispose()
...
         }
 
-        /// <inheritdoc />
-        public bool TryGetKnownCoordinate(string id, out ISpatialCoordinate spatialCoordinate)
+        bool ISpatialCoordinateService.TryGetKnownCoordinate(string id, out ISpatialCoordinate spatialCoordinate)
         {
             if (!TryParse(id, out TKey key))
             {
                 throw new ArgumentException($"Id {id} is not recognized by this coordinate service.");
             }
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/Sharing/SpatialAlignment/Common/SpatialCoordinateServiceBase.cs]: Although this was written to throw before, if we are already returning a bool, we should probably just return false here.

File: Assets/MixedRealityToolkit.Extensions/Sharing/SpatialAlignment/Common/SpatialCoordinateServiceBase.cs [hunk 3]
```diff
@@ -130,19 +135,29 @@ public virtual Task<ISpatialCoordinate> TryCreateCoordinateAsync(Vector3 worldPo
...
-        public virtual Task<bool> TryDeleteCoordinateAsync(string id, CancellationToken cancellationToken)
+        Task<bool> ISpatialCoordinateService.TryDeleteCoordinateAsync(string id, CancellationToken cancellationToken)
         {
             ThrowIfDisposed();
 
-            return Task.FromResult(false);
+            if (!TryParse(id, out TKey key))
+            {
+                throw new ArgumentException($"Id: {id} isn't valid for this spatial coordinate service.");
+            }
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/Sharing/SpatialAlignment/Common/SpatialCoordinateServiceBase.cs]: again, its probably worth returning false if this function is expected to potentially fail.

File: Assets/SpatialAlignment.ASA/SpatialAnchorsLocalizer.cs [hunk 4]
```diff
@@ -1,86 +1,86 @@
...
-    /// This is the localization mechanism for enabling anchor exchange/localization through Azure Spatial Anchors.
-    /// </summary>
-    public class SpatialAnchorsLocalizer : SpatialLocalizerBase
-    {
-#if SPATIALALIGNMENT_ASA
-        private SpatialAnchorsCoordinateService coordinateService;
-#endif
+//namespace Microsoft.MixedReality.Experimental.SpatialAlignment.AzureSpatialAnchors
+//{
+//    /// <summary>
```
Comment (chrisfromwork) [inline Assets/SpatialAlignment.ASA/SpatialAnchorsLocalizer.cs]: We should hold off on checkthing this in until asa is also insured to work.

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/MarkerDetectorLocalizationSettings.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+using System.IO;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Extensions.Experimental.SpectatorView
+{
+    public class MarkerDetectorLocalizationSettings : ISpatialLocalizationSettings
+    {
+        public int MarkerID { get; set; }
+        public float MarkerSize { get; set; }
+
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/MarkerDetectorLocalizationSettings.cs]: Can these settings also contain the marker detector type? If these settings could specify what marker detector to use or a game object prefab to create that contains an IMarkerDetector, we may be able to enable more customization?

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/MarkerVisualizerSpatialLocalizer.cs [hunk 6]
```diff
@@ -1,26 +1,26 @@
...
-        protected override Task<ISpatialCoordinate> GetHostCoordinateAsync(Guid token)
-        {
-            throw new NotImplementedException();
-        }
-    }
-}
+//        /// <inheritdoc/>
+//        protected override Task<ISpatialCoordinate> GetHostCoordinateAsync(Guid token)
+//        {
+//            throw new NotImplementedException();
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/MarkerVisualizerSpatialLocalizer.cs]: we should also comment this out with nonimplementedexceptions before checking in

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs [hunk 7]
```diff
@@ -53,188 +36,265 @@ public class SpatialCoordinateSystemManager : Singleton<SpatialCoordinateSystemM
...
-                Debug.LogWarning("Spatial localizer not specified for SpatialCoordinateSystemManager");
+                Debug.LogError($"Attempted to unregister a network manager that was not registered");
+                return;
             }
+
+            UnregisterEvents(networkManager);
         }
 
-        public void OnDisconnected(SocketEndpoint endpoint)
+        public void RestorePersistentSharedCoordinate(SocketEndpoint socketEndpoint, string sharedCoordinateName)
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs]: >RestorePersistentSharedCoordinate [](start = 20, length = 33)

nit: add this when its used compared to now.

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemParticipant.cs [hunk 8]
```diff
@@ -6,267 +6,98 @@
...
+        /// <summary>
+        /// Gets the rotation of the shared spatial coordinate in the peer device's world space.
+        /// </summary>
+        public Quaternion PeerSpatialCoordinateWorldRotation { get; internal set; }
 
-            return true;
-        }
+        public string PersistentCoordinateId { get; internal set; }
 
-        public bool TryTransformToHostWorldSpace(Quaternion localWorldRotation, out Quaternion hostWorldRotation)
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemParticipant.cs]: >TryTransformToHostWorldSpace [](start = 20, length = 28)

What facilitates this camera transform now?

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs [hunk 9]
```diff
@@ -53,188 +36,265 @@ public class SpatialCoordinateSystemManager : Singleton<SpatialCoordinateSystemM
...
             {
-                participant.Value.Dispose();
+                Debug.LogError($"Failed to find a SpatialCoordinateSystemParticipant for an attached SocketEndpoint");
+                return;
             }
 
-            participants.Clear();
+            string coordinateId = reader.ReadString();
+            //participant.PersistentCoordinateId = coordinateId;
+            //participant.Coordinate = new WorldAnchorSpatialCoordinate(coordinateId);
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs]: nit remove

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs [hunk 10]
```diff
@@ -53,188 +36,265 @@ public class SpatialCoordinateSystemManager : Singleton<SpatialCoordinateSystemM
...
+                //}
+
+                participant.IsLocatingSpatialCoordinate = true;
+                var coordinate = await currentLocalizationSession.LocalizeAsync(CancellationToken.None);
+                //if (participant.PersistentCoordinateId != null)
+                //{
+                //    participant.Coordinate = new WorldAnchorSpatialCoordinate(participant.PersistentCoordinateId, coordinate.CoordinateToWorldSpace(Vector3.zero), coordinate.CoordinateToWorldSpace(Quaternion.identity));
+                //}
+                //else
+                //{
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs]: nit: remove, readd in separate review compared to having commented out code.

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs [hunk 11]
```diff
@@ -53,188 +36,265 @@ public class SpatialCoordinateSystemManager : Singleton<SpatialCoordinateSystemM
...
+
+            GameObject participantGameObject = new GameObject($"Spatial Coordinate - {endpoint.Address}");
+            SpatialCoordinateSystemParticipant participant = participantGameObject.AddComponent<SpatialCoordinateSystemParticipant>();
+            participant.SocketEndpoint = endpoint;
+            participants[endpoint] = participant;
+
+            if (showDebugVisuals)
+            {
+                var debugVisualInstance = Instantiate(debugVisual, participant.transform);
+                debugVisualInstance.transform.localPosition = Vector3.zero;
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs]: This initially creates a valid debug visual, but we likely also want this debug visual placed correctly after any transform is applied to the camera which may take more work than just using the world position. It may actually need to be world position + accounting for camera offset

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs [hunk 12]
```diff
@@ -53,188 +36,265 @@ public class SpatialCoordinateSystemManager : Singleton<SpatialCoordinateSystemM
...
-                }
+                Debug.LogError($"Failed to find a SpatialCoordinateSystemParticipant for an attached SocketEndpoint");
+                return;
             }
+
+            participant.PeerDeviceHasTracking = reader.ReadBoolean();
+            participant.PeerSpatialCoordinateIsLocated = reader.ReadBoolean();
+            participant.PeerIsLocatingSpatialCoordinate = reader.ReadBoolean();
+            participant.PeerSpatialCoordinateWorldPosition = reader.ReadVector3();
+            participant.PeerSpatialCoordinateWorldRotation = reader.ReadQuaternion();
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs]: i think this logic should be kept in the SpatialCoordinateSystemParticipant for consistency.

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/ISpatialLocalizer.cs [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Microsoft.MixedReality.Toolkit.Extensions.Experimental.SpectatorView
+{
+    public interface ISpatialLocalizer
+    {
+        Guid SpatialLocalizerId { get; }
+
+        bool TryDeserializeSettings(BinaryReader reader, out ISpatialLocalizationSettings settings);
+
+        ISpatialLocalizationSession CreateLocalizationSession(ISpatialLocalizationSettings settings);
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/ISpatialLocalizer.cs]: >CreateLocalizationSession [](start = 36, length = 25)

This should probably be a TryCreateLocalizationSession. If a session already exists it may not create one. If the settings aren't the expected type it may fail to create one as well.

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemParticipant.cs (hunk 1 above)
Comment (matthejo) [reply Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemParticipant.cs]: This is used to determine if the device is tracking for the PeerDeviceHasTracking property. I've changed it so that this value can be platform-specific in SendCoordinateStateMessage. Might need help with getting this right for Android

Comment (matthejo) [reply Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemParticipant.cs]: This was for tracking status, that's been changed to have a platform-specific switch.

File: Assets/MixedRealityToolkit.Extensions/Sharing/SpatialAlignment/Common/SpatialCoordinateServiceBase.cs (hunk 2 above)
Comment (matthejo) [reply Assets/MixedRealityToolkit.Extensions/Sharing/SpatialAlignment/Common/SpatialCoordinateServiceBase.cs]: I think I disagree with changing this behavior. I believe passing an invalid value as an argument warrants an exception, whereas returning true/false is an indication of whether or not that particular coordinate exists.

File: Assets/MixedRealityToolkit.Extensions/Sharing/SpatialAlignment/Common/SpatialCoordinateServiceBase.cs (hunk 3 above)
Comment (matthejo) [reply Assets/MixedRealityToolkit.Extensions/Sharing/SpatialAlignment/Common/SpatialCoordinateServiceBase.cs]: Same as the other comment, I think invalid arguments warrants an exception (as a programmer error) compared to an expected failure to find something.

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs (hunk 7 above)
Comment (matthejo) [reply Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs]: Removed this for now

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/MarkerDetectorLocalizationSettings.cs (hunk 5 above)
Comment (chrisfromwork) [reply Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/MarkerDetectorLocalizationSettings.cs]: this seems beyond whats needed for now.
